specifications/unified-provider.org

```diff
@@ -28,57 +42,197 @@ Multipass) and potential future providers (Remote LXD, etc.).
+
+      :param step: step to load environment for.
+      :param part_name: part_name to load environment for.  If step is specified,
+      but not a part_name.
+
+      :raises SnapcraftProviderUnsupportedMethod: if provider does not support
+      providing a shell.
+      """
+
+      ...
+#+end_src
+
+*** Summary of changes against current provider:
+- =is_provider_ready()= replaces classmethod =ensure_provider()=.
+- =setup_provider()= becomes an instance method, supporting non-interactive
+  configuration, if possible.
+- Removes =mount_project()= interface. Provider will mount project as required.
 - Replace =pack_project()= with =snap()=, adding the implication that build steps
-  are to be executed as required (and packing may not be done by Snapcraft).
+  are to be executed as required, removing the implication that Snapcraft itself
+  invokes the packing (Snapcraft may just fetch the output of the provider).
 - Rename =clean()= to =clean_parts()=, to improve clarity of scope.
 - Rename =clean_project()= to =clean()=, to improve clarity of scope.
+- Context manager =__enter__()= and =__exit__()= are currently used to define
+  the lifetime start/stop status of containers. =__enter__()= will no longer
+  start an instance, but =_exit()= will continue to stop the instance, if
+  configured (and applicable) to do so.
 
-** Migrate current logic into new providers:
-- HostProvider
-- ManagedHostProvider, inherits much of HostProvider.
-- LaunchpadProvider
-- RemoteLXDProvider, inherits much of LXDProvider, overrides the (now internal)
-  equivalent of mount_project() behavior with a rsync from host to remote.
-
-* CLI Handling
+* CLI Refactoring
 
-** Provider NotImplementedError
+** SnapcraftProviderUnsupportedMethodError
 
 Certain commands will not supported by all providers. E.g.
-- =snapcraft try= is not compatible with Launchpad, Remote LXD, Host
-- =snapcraft <step>= is not compatible with Lauchpad, Remote LXD
-- =snapcraft clean <parts>= is not compatible with Launchpad, Remote LXD
+- =snapcraft <step>= is not compatible with Lauchpad.
+- =snapcraft clean <parts>= is not compatible with Launchpad
 
-In these cases, the applicable CLI methods will catch the NotImplementedError
-and inform the user that this command is not available with the current
-configured provider, e.g.:
+In these cases, the applicable CLI methods will catch the
+=SnapcraftProviderUnsupportedMethodError= and inform the user that this command
+is not available with the current configured provider.
 
-=Sorry, '<command>' is not available for the currently configured provider
+*** Lifecycle command =snapcraft <step> ...=
+
+If attempting to run a lifecycle command (other than =snap=) and
+the provider does not support it, fail with the following error:
+='<command>' is not available for the currently configured provider
 '<provider>'.=
 
-For =snapcraft clean <parts>=:
+For =--debug=, =--shell=, and =--shell-after= options:
```

Comment:
should these be enforced by Provider sanity checks and not the cli implementation? It is not clarified here.

Reply:
I was planning on cli and having the cli handle the exception that it's not supported.  However I just realized some operations may run before it hits the exception. Perhaps it'd be best to have the providers advertise the features they support, or add to provider sanity checks.  I'll consider it further

Reply:
I would prefer that the "capabilties" be described by the provider and the cli to raise an exception

Reply:
:+1: